compiler/runtime.go

```diff
@@ -418,103 +449,112 @@ func (e *unaryEvaluator) Eval(ctx context.Context, scope Scope) (values.Value, e
 	if err != nil {
 		return nil, err
 	}
+	ret.Retain()
 	return ret, nil
 }
 
 type integerEvaluator struct {
-	i int64
+	v values.Value
 }
 
 func (e *integerEvaluator) Type() semantic.MonoType {
 	return semantic.BasicInt
 }
 
 func (e *integerEvaluator) Eval(ctx context.Context, scope Scope) (values.Value, error) {
-	return values.NewInt(e.i), nil
+	e.v.Retain()
+	return e.v, nil
 }
 
 type unsignedIntegerEvaluator struct {
-	i uint64
+	v values.Value
 }
 
 func (e *unsignedIntegerEvaluator) Type() semantic.MonoType {
 	return semantic.BasicUint
 }
 
 func (e *unsignedIntegerEvaluator) Eval(ctx context.Context, scope Scope) (values.Value, error) {
-	return values.NewUInt(e.i), nil
+	e.v.Retain()
+	return e.v, nil
 }
 
 type stringEvaluator struct {
-	s string
+	v values.Value
 }
 
 func (e *stringEvaluator) Type() semantic.MonoType {
 	return semantic.BasicString
 }
 
 func (e *stringEvaluator) Eval(ctx context.Context, scope Scope) (values.Value, error) {
-	return values.NewString(e.s), nil
+	e.v.Retain()
+	return e.v, nil
 }
 
 type regexpEvaluator struct {
-	r *regexp.Regexp
+	v values.Value
 }
 
 func (e *regexpEvaluator) Type() semantic.MonoType {
 	return semantic.BasicRegexp
 }
 
 func (e *regexpEvaluator) Eval(ctx context.Context, scope Scope) (values.Value, error) {
-	return values.NewRegexp(e.r), nil
+	e.v.Retain()
+	return e.v, nil
 }
 
 type booleanEvaluator struct {
-	b bool
+	v values.Value
```

Comment:
Seeing the change from a native type to an interface makes me wonder about any performance impact to non-vectorized functions. What's your take on this?

Also, do we need to call `Retain` here? This change doesn't seem necessary for the current limited case that is handled by vectorization. When we do handle something like `(r) => ({r with a: 10.0})` I would expect that the vectorization pass would insert a new operation `(r) => ({r with a: vector_from_scalar(10.0)})`. Meaning even in that case, we would never have a (say) `booleanEvaluator` that produces a vector that needs to be reference counted.

(We are calling `vector_from_scalar` something different in the issues that were filed. The name we landed on is escaping me at the moment.)

Thoughts?